File: home/android/app/src/main/java/page/bento/home/Releases.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 The Bento authors

package page.bento.home

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.File
import java.math.BigInteger
import java.net.HttpURLConnection
import java.net.URL
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import java.util.Base64

/**
 * Where a NEW document comes from.
 *
 * **Starter decks are never bundled.** They change often, and there are already
 * three apps with more coming — so a bundled seed means either picking one
 * arbitrarily or shipping several copies of Bento inside the app, each stale
 * from the moment it was built. Measured before this changed: the single
 * bundled slides seed was 517,161 bytes, **81% of a 630,851-byte release APK**.
 *
 * So the shell is fetched from the same signed release channel a document uses
 * to update itself, which also means a document created here is the version
 * everyone else has, the same day.
 *
 * THIS IS THE ONE THING THE HOST FETCHES. The app makes no other network request
 * — no update check of its own, no telemetry. `docs/PLATFORM.md` §1 requires no
 * network to OPEN, EDIT, PRESENT or SAVE, and none of those touch this; creating
 * a document from a template is a different act, and `home/webext` draws the
 * line in the same place. The result is cached, so only the first "New" of a
 * given release needs a connection.
 *
 * VERIFICATION IS NOT OPTIONAL HERE. The bytes become an executable HTML
 * document on the user's own disk that they will afterwards trust, so an
 * unverified download lets a network attacker choose what they create. Both
 * halves are checked, exactly as `kernel/src/update.ts` does it: the manifest's
 * ECDSA signature, and then the shell's sha256 as pinned by the signed payload.
 * (`home/webext`'s equivalent does neither — raised separately.)
 */
object Releases {

    private const val TAG = "BentoTray"

    /** The apps a new document can be. Mirrors `APPS` in
     *  `home/webext/src/library.js`; adding one here is the whole integration,
     *  because nothing else in this app asks which Bento a document is. */
    data class App(val id: String, val label: String, val blurb: String) {
        val manifest get() = "https://bento.page/releases/$id/manifest.json"
        /** What a manifest for this app must call itself. Compared against what
         *  the CALLER asked for, never against what the payload claims. */
        val appId get() = "bento-$id"
    }

    val APPS = listOf(
        App("slides", "Slides", "Presentations"),
        App("spaces", "Spaces", "Notes and pages"),
        App("dash", "Dash", "Data and sheets"),
    )

    /** Release signing PUBLIC key, copied from `kernel/src/update.ts`. The
     *  private half lives offline with the maintainer. Rotating it orphans every
     *  previously shipped file, so the key is guarded rather than rotated. */
    private const val KEY_X = "GMHSKwWcAoJVq-Dz1ZxWZM6TXATWIKbaQBpjoTystH8"
    private const val KEY_Y = "flFNzbdXCmJN8RQYCeG71rBZnnbN-MCEnp1EbCLFrj0"

    private const val TIMEOUT = 20_000

    class Failed(message: String) : Exception(message)

    /**
     * The current signed shell for [app], from cache when we already have it.
     *
     * Blocking — call it off the main thread.
     */
    fun seedFor(c: Context, app: App): ByteArray =
        resolve(app, Deps(
            get = ::get,
            readFloor = { key -> floors(c).getString(key, null) },
            writeFloor = { key, v -> floors(c).edit().putString(key, v).apply() },
            cached = { v -> cached(c, app, v) },
            store = { v, b -> put(c, app, v, b) },
        ))

    /**
     * Everything the network can lie about, in the order it has to be checked.
     *
     * Dependencies are injected so the sequence can be tested on the JVM without
     * a device — the two cases that matter (a high version whose bytes fail, and
     * an equal version) are exactly the ones nobody exercises by hand.
     */
    internal class Deps(
        val get: (String) -> ByteArray,
        val readFloor: (String) -> String?,
        val writeFloor: (String, String) -> Unit,
        val cached: (String) -> ByteArray? = { null },
        val store: (String, ByteArray) -> Unit = { _, _ -> },
    )

    internal fun resolve(app: App, d: Deps): ByteArray {
        // Only slides has a published channel today; spaces and dash 404. Say so
        // rather than surfacing an HTTP code — the app list is aspirational on
        // purpose, so an unreleased one is an expected answer, not a fault.
        val envelope = try { d.get(app.manifest) } catch (e: Failed) {
            if (e.message?.contains("404") == true)
                throw Failed("${app.label} has not been released yet")
            throw e
        }
        // The signature is checked HERE, over the exact bytes that arrived, and
        // everything after it works from the verified payload. Split so the
        // ORDER of the remaining checks can be tested on the JVM without
        // manufacturing a signing key — the sequence is what gets these wrong,
        // not the crypto.
        val payload = verify(String(envelope, Charsets.UTF_8))
        return resolveVerified(app, parse(payload), d)
    }

    /**
     * What a verified payload says, as data.
     *
     * The seam is a struct rather than a JSON string on purpose: `org.json` is
     * an Android STUB on the JVM rig, so a seam that parses there would be
     * testing the stub. Parsing happens once, on the device side of the line;
     * the ORDER of the checks — which is what gets these wrong — is then pure.
     */
    internal data class Release(
        val app: String?, val version: String, val url: String, val sha256: String,
    )

    private fun parse(payload: JSONObject): Release = Release(
        // "" for an absent field, normalised to null so the identity check
        // below compares a genuine absence rather than an empty string.
        app = payload.optString("app").ifEmpty { null },
        version = payload.optString("version").ifEmpty {
            throw Failed("the release server did not name a version")
        },
        url = payload.optString("url").ifEmpty {
            throw Failed("the release server did not offer a build")
        },
        sha256 = payload.optString("sha256").lowercase().ifEmpty {
            throw Failed("the release is not pinned to a hash")
        },
    )

    internal fun resolveVerified(app: App, r: Release, d: Deps): ByteArray {
        // IDENTITY. A genuinely signed bento-slides manifest served on the dash
        // channel passes the signature AND the digest — every byte authentic,
        // just not what was asked for. Compared against the app the CALLER
        // wanted, never against what the payload says about itself.
        //
        // An ABSENT field must not read as a match. `named` is null when the
        // field is missing and `want` is never null, so the comparison already
        // refuses it — tidying this to `if (named != null && named != want)`
        // would silently turn a missing field into a pass, which is the hole
        // home/webext had.
        val want = app.appId
        val named = r.app
        if (named != want) {
            throw Failed("the release channel offered ${named ?: "an unnamed app"}, not $want — refusing it")
        }

        val version = r.version

        // NO DOWNGRADES. Everything above passes for a replayed OLD release: it
        // is genuinely signed, it names the right app, and its shell really does
        // hash to its pin. Every byte authentic — just last month's. That is
        // what survives an origin or CDN compromise where the attacker can
        // re-serve but cannot forge, so remembering is the only thing that
        // catches it. kernel/src/update.ts refuses to go backwards too, but it
        // has its own build version to compare against; a host that CREATES
        // documents has nothing to compare against but what it has seen.
        val floor = readFloor(app, d)
        if (floor != null && compareVersions(version, floor) < 0) {
            throw Failed(
                "the ${app.label} channel offered $version, older than the $floor " +
                    "already seen — refusing it"
            )
        }

        d.cached(version)?.let { return it }

        val shell = d.get(r.url)
        val got = MessageDigest.getInstance("SHA-256").digest(shell)
            .joinToString("") { "%02x".format(it) }
        if (got != r.sha256) throw Failed("the downloaded app did not match its signed hash — refusing it")

        // Raised only NOW, after the bytes passed their digest. Raising it on a
        // merely-verified manifest would let one forged-but-unfetchable release
        // naming 9.9.9 lock this device out of every real release below it —
        // turning a failed attack into a permanent denial of service.
        raiseFloor(app, version, floor, d)

        d.store(version, shell)
        return shell
    }

    // ------------------------------------------------------------------- floor

    private fun floorKey(app: App) = "release-floor:${app.id}"

    /** An unreadable store reads as NO FLOOR, not as a refusal. Private mode,
     *  quota, a migration mid-flight — being unable to remember must not mean
     *  being unable to create a document. */
    private fun readFloor(app: App, d: Deps): String? =
        try { d.readFloor(floorKey(app))?.ifEmpty { null } } catch (_: Exception) { null }

    private fun raiseFloor(app: App, version: String, floor: String?, d: Deps) {
        // EQUAL is not a downgrade. Re-fetching the version already seen is the
        // normal case — the second document somebody creates — so treating it as
        // a rollback would break the + button on its SECOND USE rather than at
        // some exotic edge.
        if (floor != null && compareVersions(version, floor) <= 0) return
        // Best effort: a floor that cannot be written costs protection next
        // time, not this document.
        try { d.writeFloor(floorKey(app), version) } catch (_: Exception) { }
    }

    /**
     * The kernel's comparison semantics, not a third set.
     *
     * Note what JavaScript's `(pa[i] || 0) - (pb[i] || 0)` actually does with a
     * component that will not parse: `Number('1a')` is NaN, NaN is FALSY, so it
     * coerces to 0 rather than propagating. A strange version can therefore fail
     * to RAISE the floor but can never block a release. `toIntOrNull() ?: 0`
     * reproduces that exactly.
     */
    internal fun compareVersions(a: String, b: String): Int {
        val pa = a.split(".")
        val pb = b.split(".")
        for (i in 0 until maxOf(pa.size, pb.size)) {
            val d = (pa.getOrNull(i)?.toIntOrNull() ?: 0) - (pb.getOrNull(i)?.toIntOrNull() ?: 0)
            if (d != 0) return d
        }
        return 0
    }

    // ------------------------------------------------------------------ crypto

    /**
     * Verify the `{payload, sig}` envelope and return the parsed payload.
     *
     * The signature covers the EXACT payload string bytes — no canonicalisation,
     * no re-serialisation. Parse it for reading only, never to re-encode and
     * verify, or a byte-identical-looking round trip breaks the signature.
     */
    private fun verify(raw: String): JSONObject {
        val env = try { JSONObject(raw) } catch (e: Exception) {
            throw Failed("the release manifest is not valid JSON")
        }
        val payload = env.optString("payload")
        val sig = env.optString("sig")
        if (payload.isEmpty() || sig.isEmpty()) throw Failed("the release manifest is malformed")

        val params = AlgorithmParameters.getInstance("EC").apply {
            init(ECGenParameterSpec("secp256r1"))
        }.getParameterSpec(ECParameterSpec::class.java)

        val point = ECPoint(BigInteger(1, b64url(KEY_X)), BigInteger(1, b64url(KEY_Y)))
        val key = KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(point, params))

        val ok = try {
            Signature.getInstance("SHA256withECDSA").run {
                initVerify(key)
                update(payload.toByteArray(Charsets.UTF_8))
                verify(derOf(Base64.getMimeDecoder().decode(sig)))
            }
        } catch (e: Exception) {
            Log.w(TAG, "signature check errored", e); false
        }
        if (!ok) throw Failed("the release manifest signature is INVALID — refusing it")

        return try { JSONObject(payload) } catch (e: Exception) {
            throw Failed("the release payload is not valid JSON")
        }
    }

    /**
     * WebCrypto emits ECDSA signatures as RAW `r || s`; Java expects DER.
     *
     * This is the whole reason a correct-looking verification can fail: the
     * manifest is signed by `scripts/sign-release.mjs` through WebCrypto, so its
     * 64 bytes are two fixed-width integers, and handing those to
     * `SHA256withECDSA` unconverted is simply a bad signature. Silent, and
     * indistinguishable from tampering.
     */
    private fun derOf(raw: ByteArray): ByteArray {
        if (raw.size != 64) return raw   // already DER, or not something we made
        fun int(b: ByteArray): ByteArray {
            var i = 0
            while (i < b.size - 1 && b[i] == 0.toByte()) i++          // drop leading zeros
            var v = b.copyOfRange(i, b.size)
            if (v[0].toInt() and 0x80 != 0) v = byteArrayOf(0) + v    // keep it positive
            return byteArrayOf(0x02, v.size.toByte()) + v
        }
        val body = int(raw.copyOfRange(0, 32)) + int(raw.copyOfRange(32, 64))
        return byteArrayOf(0x30, body.size.toByte()) + body
    }

    /** java.util.Base64, not android.util.Base64: the Android one is a STUB in
     *  unit tests, and this routine is the part most worth testing. Available
     *  since API 26, which is minSdk. */
    private fun b64url(s: String): ByteArray = Base64.getUrlDecoder().decode(s)

    // ------------------------------------------------------------------- net

    private fun get(url: String): ByteArray {
        if (!url.startsWith("https://")) throw Failed("the release server offered a non-HTTPS build")
        val c = URL(url).openConnection() as HttpURLConnection
        c.connectTimeout = TIMEOUT
        c.readTimeout = TIMEOUT
        c.setRequestProperty("Cache-Control", "no-store")
        // ASK FOR BYTES, NOT A PAGE. HttpURLConnection's default Accept is
        // browser-shaped ("text/html, image/gif, …"), and bento.page sits behind
        // a CDN that INJECTS a tracking beacon into anything it believes is a
        // page being browsed. That made the download 359 bytes longer than the
        // release the maintainer signed, and the hash check below refused it —
        // correctly. `Accept: */*` asks for the artifact itself.
        //
        // The hash check is the actual defence and stays regardless: this header
        // avoids a known rewrite, it does not make the bytes trustworthy.
        c.setRequestProperty("Accept", "*/*")
        try {
            if (c.responseCode !in 200..299) throw Failed("the release server answered ${c.responseCode}")
            return c.inputStream.use { it.readBytes() }
        } finally {
            c.disconnect()
        }
    }

    // ----------------------------------------------------------------- cache

    /** Kept in filesDir, not cacheDir: this is what makes "New" work offline
     *  after the first time, so it should not evaporate under storage pressure
     *  the way a thumbnail happily can. */
    private fun floors(c: Context) =
        c.getSharedPreferences("releases", Context.MODE_PRIVATE)

    private fun cacheFile(c: Context, app: App, version: String) =
        File(File(c.filesDir, "seeds"), "${app.id}-$version.html")

    private fun cached(c: Context, app: App, version: String): ByteArray? {
        val f = cacheFile(c, app, version)
        return if (f.exists() && f.length() > 0) try { f.readBytes() } catch (_: Exception) { null }
        else null
    }

    private fun put(c: Context, app: App, version: String, bytes: ByteArray) {
        try {
            val f = cacheFile(c, app, version)
            f.parentFile?.mkdirs()
            f.writeBytes(bytes)
            // One release per app is enough to work offline; older ones are dead
            // weight the moment a newer one is cached.
            f.parentFile?.listFiles()
                ?.filter { it.name.startsWith("${app.id}-") && it.name != f.name }
                ?.forEach { it.delete() }
        } catch (e: Exception) {
            Log.w(TAG, "could not cache the seed", e)
        }
    }

    /** The newest cached shell for [app], whatever its version — the offline
     *  fallback when the release server cannot be reached at all. */
    fun anyCached(c: Context, app: App): ByteArray? =
        File(c.filesDir, "seeds").listFiles()
            ?.filter { it.name.startsWith("${app.id}-") }
            ?.maxByOrNull { it.lastModified() }
            ?.let { try { it.readBytes() } catch (_: Exception) { null } }
}

File: home/ios/Releases.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 The Bento authors

import CryptoKit
import Foundation

/// Where a NEW document comes from.
///
/// **Starter decks are never bundled.** They change often, and there are already
/// three apps with more coming — so a bundled seed means either picking one
/// arbitrarily or shipping several copies of Bento inside the app, each stale
/// from the moment it was built. Measured on the Android host before this
/// changed: the single bundled slides seed was 517,161 bytes, **81% of a
/// 630,851-byte release APK**. The `.ipa` was never measured, but it staged the
/// same 686KB shell.
///
/// So the shell is fetched from the same signed release channel a document uses
/// to update itself, which also means a document created here is the version
/// everyone else has, the same day.
///
/// THIS IS THE ONE THING THE HOST FETCHES. The app makes no other network
/// request — no update check of its own, no telemetry. `docs/PLATFORM.md` §1
/// requires no network to OPEN, EDIT, PRESENT or SAVE, and none of those touch
/// this; creating a document from a template is a different act, and
/// `home/webext` draws the line in the same place. The result is cached, so only
/// the first "New" of a given release needs a connection.
///
/// VERIFICATION IS NOT OPTIONAL HERE. The bytes become an executable HTML
/// document on the user's own disk that they will afterwards trust, so an
/// unverified download lets a network attacker choose what they create. Both
/// halves are checked, exactly as `kernel/src/update.ts` does it: the manifest's
/// ECDSA signature, and then the shell's sha256 as pinned by the signed payload.
///
/// Deliberately a mirror of `home/android`'s `Releases.kt` — same apps, same
/// key, same order of checks, same refusal messages — so the two hosts cannot
/// answer differently. Two details there were learned the hard way and are
/// inherited rather than rediscovered: the Accept header (see `fetch`), and the
/// signature being raw `r || s` rather than DER (see `verify`).
enum Releases {

    /// The apps a new document can be. Mirrors `APPS` in
    /// `home/webext/src/library.js` and `Releases.kt`; adding one here is the
    /// whole integration, because nothing else in this app asks which Bento a
    /// document is.
    struct App: Equatable {
        let id: String
        let label: String
        let blurb: String
        var manifest: String { "https://bento.page/releases/\(id)/manifest.json" }
    }

    static let apps = [
        App(id: "slides", label: "Slides", blurb: "Presentations"),
        App(id: "spaces", label: "Spaces", blurb: "Notes and pages"),
        App(id: "dash", label: "Dash", blurb: "Data and sheets"),
    ]

    /// Release signing PUBLIC key, copied from `kernel/src/update.ts`. The
    /// private half lives offline with the maintainer. Rotating it orphans every
    /// previously shipped file, so the key is guarded rather than rotated.
    private static let keyX = "GMHSKwWcAoJVq-Dz1ZxWZM6TXATWIKbaQBpjoTystH8"
    private static let keyY = "flFNzbdXCmJN8RQYCeG71rBZnnbN-MCEnp1EbCLFrj0"

    private static let timeout: TimeInterval = 20

    struct Failed: LocalizedError {
        let message: String
        var errorDescription: String? { message }
    }

    // MARK: - The release

    struct Release {
        let version: String
        let url: String
        let sha256: String
    }

    /// Verify an envelope and pull out what a download needs — including that it
    /// is a release of the app we ASKED for.
    ///
    /// That last check is not redundant with the signature, and it was missing
    /// here until `home/webext`'s own verification (PR #318) tested for it. Both
    /// a `bento-slides` manifest and a `bento-dash` one are genuinely signed by
    /// the same maintainer key, so serving the former on the latter's channel
    /// passes the signature AND the hash: every byte is authentic, just not the
    /// thing that was requested. Only identity catches a swap between two real
    /// releases. `home/android` does not check this either — reported.
    static func release(from raw: String, for app: App, notBefore floor: String? = nil) throws -> Release {
        let payload = try verify(raw)

        // The payload names the app as `bento-<id>`.
        let want = "bento-\(app.id)"
        guard let named = payload["app"] as? String, named == want else {
            let got = (payload["app"] as? String) ?? "nothing"
            throw Failed(message: "the \(app.label) channel served a release for \(got) — refusing it")
        }
        guard let version = payload["version"] as? String, !version.isEmpty else {
            throw Failed(message: "the release server did not name a version")
        }
        guard let url = payload["url"] as? String, !url.isEmpty else {
            throw Failed(message: "the release server did not offer a build")
        }
        guard let sha = (payload["sha256"] as? String)?.lowercased(), !sha.isEmpty else {
            throw Failed(message: "the release is not pinned to a hash")
        }
        // ROLLBACK REPLAY. A stale but GENUINE manifest passes every check above
        // — signature, app identity, and the digest of the shell it points at,
        // because all of it really was signed and really does match. It just
        // hands over an older release, which is how someone who can re-serve but
        // not forge pins new documents to a version with a known hole in it.
        //
        // `kernel/src/update.ts` already refuses to go backwards; a host that
        // CREATES documents had no such floor, because it has no version of its
        // own to compare against. The floor is therefore the highest version
        // this device has already accepted for this app.
        //
        // The cost is honest and worth stating: a deliberate rollback by the
        // maintainer — pulling a bad release — is refused too, until the version
        // number moves past it. That is the same trade update.ts makes, so at
        // least the two agree.
        if let floor, isOlder(version, than: floor) {
            throw Failed(message: "the \(app.label) channel offered \(version), older than the \(floor) "
                       + "this device already accepted — refusing it")
        }
        return Release(version: version, url: url, sha256: sha)
    }

    /// Dotted-numeric compare, enough for the versions this project mints. An
    /// unparsable component sorts as 0 rather than throwing: a strange version
    /// string should not be able to BLOCK an update, only fail to raise the floor.
    static func isOlder(_ a: String, than b: String) -> Bool {
        let x = a.split(separator: ".").map { Int($0.prefix(while: \.isNumber)) ?? 0 }
        let y = b.split(separator: ".").map { Int($0.prefix(while: \.isNumber)) ?? 0 }
        for i in 0..<max(x.count, y.count) {
            let l = i < x.count ? x[i] : 0
            let r = i < y.count ? y[i] : 0
            if l != r { return l < r }
        }
        return false
    }

    /// The highest version this device has accepted for `app`, if any.
    static func floor(for app: App) -> String? {
        UserDefaults.standard.string(forKey: "bento.tray.release-floor.\(app.id)")
    }

    private static func raiseFloor(_ app: App, to version: String) {
        if let current = floor(for: app), !isOlder(current, than: version) { return }
        UserDefaults.standard.set(version, forKey: "bento.tray.release-floor.\(app.id)")
    }

    // MARK: - The seed

    /// The current signed shell for `app`, from cache when we already have it.
    static func seed(for app: App) async throws -> Data {
        // Only slides has a published channel today; spaces and dash 404. Say so
        // rather than surfacing an HTTP code — the app list is aspirational on
        // purpose (adding an app here is the whole integration), so an unreleased
        // one is an expected answer, not a fault.
        let envelope: Data
        do {
            envelope = try await fetch(app.manifest)
        } catch let e as Failed where e.message.contains("404") {
            throw Failed(message: "\(app.label) has not been released yet")
        }

        let release = try release(from: String(decoding: envelope, as: UTF8.self),
                                  for: app, notBefore: floor(for: app))

        if let hit = cached(app, release.version) {
            raiseFloor(app, to: release.version)
            return hit
        }

        let shell = try await fetch(release.url)
        let got = SHA256.hash(data: shell).map { String(format: "%02x", $0) }.joined()
        // The signed payload pins this. A mismatch means the bytes are not the
        // release the maintainer signed, whatever the server said.
        guard got == release.sha256 else {
            throw Failed(message: "the downloaded app did not match its signed hash — refusing it")
        }

        // Only after the bytes have proven themselves. Raising the floor on a
        // manifest whose download then failed its hash would let a forged
        // manifest lock the device out of every real release below it.
        store(app, release.version, shell)
        raiseFloor(app, to: release.version)
        return shell
    }

    // MARK: - Crypto

    /// Verify the `{payload, sig}` envelope and return the parsed payload.
    ///
    /// The signature covers the EXACT payload string bytes — no canonicalisation,
    /// no re-serialisation. Parse it for reading only, never to re-encode and
    /// verify, or a byte-identical-looking round trip breaks the signature.
    ///
    /// `scripts/sign-release.mjs` signs through WebCrypto, which emits ECDSA
    /// signatures as RAW `r || s` — 64 bytes, two fixed-width integers.
    /// CryptoKit's `ECDSASignature(rawRepresentation:)` takes exactly that, so
    /// iOS needs no conversion. Noted because the Android host DOES: Java's
    /// `SHA256withECDSA` wants DER, and handing it the raw 64 bytes is simply a
    /// bad signature — silent, and indistinguishable from tampering. Anyone
    /// porting this a third time should check which their platform expects
    /// before concluding the manifest is broken.
    /// Internal rather than private so `scripts/test-tray-releases.mjs` can prove
    /// it REFUSES things. A verifier that has only ever been watched saying yes
    /// is indistinguishable from `return true`.
    static func verify(_ raw: String) throws -> [String: Any] {
        guard let data = raw.data(using: .utf8),
              let env = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { throw Failed(message: "the release manifest is not valid JSON") }

        let payloadField = env["payload"] as? String
        let sigField = env["sig"] as? String
        // An UNSIGNED manifest is refused as its own category, not as a parse
        // error. `{"url": …}` flat is exactly the shape the old broken reader
        // was reaching for, so saying "malformed" invites someone later to add a
        // lenient fallback for it as a compatibility gap. It is not a gap.
        if (payloadField ?? "").isEmpty && (sigField ?? "").isEmpty {
            throw Failed(message: "the release manifest is not signed — refusing it")
        }
        guard let payload = payloadField, !payload.isEmpty,
              let sig = sigField, !sig.isEmpty
        else { throw Failed(message: "the release manifest is malformed") }

        guard let x = b64url(keyX), let y = b64url(keyY),
              let key = try? P256.Signing.PublicKey(rawRepresentation: x + y)
        else { throw Failed(message: "the embedded release key is unusable") }

        guard let sigBytes = Data(base64Encoded: sig) ?? b64url(sig),
              let signature = try? P256.Signing.ECDSASignature(rawRepresentation: sigBytes),
              key.isValidSignature(signature, for: Data(payload.utf8))
        else { throw Failed(message: "the release manifest signature is INVALID — refusing it") }

        guard let parsed = try? JSONSerialization.jsonObject(with: Data(payload.utf8)) as? [String: Any]
        else { throw Failed(message: "the release payload is not valid JSON") }
        return parsed
    }

    private static func b64url(_ s: String) -> Data? {
        var t = s.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        while t.count % 4 != 0 { t += "=" }
        return Data(base64Encoded: t)
    }

    // MARK: - Net

    private static func fetch(_ url: String) async throws -> Data {
        guard url.hasPrefix("https://"), let u = URL(string: url) else {
            throw Failed(message: "the release server offered a non-HTTPS build")
        }
        var req = URLRequest(url: u, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
                             timeoutInterval: timeout)
        // ASK FOR BYTES, NOT A PAGE — and this header is LOAD-BEARING, not
        // belt-and-braces. Measured against the live server 2026-08-17:
        //
        //   Accept: */*                       689,316 bytes — matches the pin
        //   Accept: text/html,…,*/*;q=0.8     689,675 bytes — does NOT
        //
        // Same URL, same `.bento.html` path, 359 bytes apart. The CDN injects a
        // Cloudflare analytics beacon before `</body>` when it believes the
        // response is a page being browsed, and the trigger is the ACCEPT
        // HEADER, not the extension — which is worth knowing precisely, because
        // if it were the path then no header would fix it for anyone.
        //
        // The injected file still carries an intact `id="bento-doc"`, so it
        // looks like a perfectly good document; only the hash tells them apart.
        // (An earlier note here said this had been fixed at the origin. It has
        // not — that was inherited second-hand and is wrong.)
        //
        // The hash check remains the actual defence: this header avoids a known
        // rewrite, it does not make the bytes trustworthy.
        req.setValue("*/*", forHTTPHeaderField: "Accept")
        req.setValue("no-store", forHTTPHeaderField: "Cache-Control")

        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw Failed(message: "the release server gave no answer")
        }
        guard (200...299).contains(http.statusCode) else {
            throw Failed(message: "the release server answered \(http.statusCode)")
        }
        return data
    }

    // MARK: - Cache

    /// Kept in Application Support, not Caches: this is what makes "New" work
    /// offline after the first time, so it should not evaporate under storage
    /// pressure the way a thumbnail happily can.
    private static func seedsDir() -> URL? {
        guard let dir = FileManager.default.urls(for: .applicationSupportDirectory,
                                                 in: .userDomainMask).first?
            .appendingPathComponent("seeds") else { return nil }
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private static func cacheFile(_ app: App, _ version: String) -> URL? {
        seedsDir()?.appendingPathComponent("\(app.id)-\(version).html")
    }

    private static func cached(_ app: App, _ version: String) -> Data? {
        guard let f = cacheFile(app, version), let d = try? Data(contentsOf: f), !d.isEmpty
        else { return nil }
        return d
    }

    private static func store(_ app: App, _ version: String, _ bytes: Data) {
        guard let f = cacheFile(app, version), let dir = seedsDir() else { return }
        do { try bytes.write(to: f, options: .atomic) } catch { return }
        // One release per app is enough to work offline; older ones are dead
        // weight the moment a newer one is cached.
        let others = (try? FileManager.default.contentsOfDirectory(at: dir,
                                                                   includingPropertiesForKeys: nil)) ?? []
        for other in others where other.lastPathComponent.hasPrefix("\(app.id)-")
            && other.lastPathComponent != f.lastPathComponent {
            try? FileManager.default.removeItem(at: other)
        }
    }

    /// The newest cached shell for `app`, whatever its version — the offline
    /// fallback when the release server cannot be reached at all. Sound rather
    /// than a guess: it was verified when it was cached.
    static func anyCached(_ app: App) -> Data? {
        guard let dir = seedsDir(),
              let files = try? FileManager.default.contentsOfDirectory(
                at: dir, includingPropertiesForKeys: [.contentModificationDateKey])
        else { return nil }
        let mine = files.filter { $0.lastPathComponent.hasPrefix("\(app.id)-") }
        let newest = mine.max { a, b in
            let da = (try? a.resourceValues(forKeys: [.contentModificationDateKey]))?
                .contentModificationDate ?? .distantPast
            let db = (try? b.resourceValues(forKeys: [.contentModificationDateKey]))?
                .contentModificationDate ?? .distantPast
            return da < db
        }
        return newest.flatMap { try? Data(contentsOf: $0) }
    }
}
